### PR TITLE
[MXNET-175] Raise user warning on mismatch between module data_names and data iter names

### PR DESCRIPTION
**Pull request by spidyDev**
## Description ##
dataiter data_names  should match module data names.
Raising a warning if it does not match.

Related github issue:
https://github.com/apache/incubator-mxnet/issues/10045

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

@sandeep-krishnamurthy @nswamy @lupesko @anirudh2290

### CODE REVIEW

File: python/mxnet/module/base_module.py [hunk 1]
```diff
@@ -363,6 +363,19 @@ def predict(self, eval_data, num_batch=None, merge_batches=True, reset=True,
             eval_data.reset()
 
         output_list = []
+        # If data provided is a dict. This makes sure that only relevant data items
```
Comment (anirudh2290) [inline python/mxnet/module/base_module.py]: Do we need to make this change for score too ?

File: python/mxnet/io.py [hunk 2]
```diff
@@ -645,8 +647,8 @@ def __init__(self, data, label=None, batch_size=1, shuffle=False,
                  label_name='softmax_label'):
         super(NDArrayIter, self).__init__(batch_size)
 
-        self.data = _init_data(data, allow_empty=False, default_name=data_name)
-        self.label = _init_data(label, allow_empty=True, default_name=label_name)
+        self.data, self.renamedData = _init_data(data, allow_empty=False, default_name=data_name)
```
Comment (anirudh2290) [inline python/mxnet/io.py]: snake case for renamedData?

Comment (spidyDev) [reply python/mxnet/io.py]: done !

File: python/mxnet/module/base_module.py (hunk 1 above)
Comment (spidyDev) [reply python/mxnet/module/base_module.py]: Yes, thanks for pointing this out. Added to score() too!!

**Comment by anirudh2290** [comment]
@reminisce can you help take a look ?

File: python/mxnet/module/base_module.py [hunk 3]
```diff
@@ -364,6 +385,22 @@ def predict(self, eval_data, num_batch=None, merge_batches=True, reset=True,
...
+                eval_data.data and \
+                isinstance(eval_data.data[0], tuple):
+            # Keeping a copy of original data.
+            orig_eval_data = eval_data.data
+            data_dict = dict(eval_data.data)
+
+            for k, _ in eval_data.data:
+                if k not in self.data_names:
+                    del data_dict[k]
+            eval_data.data = list(data_dict.items())
```
[the remark this answers is not in the document]
Comment (spidyDev) [reply python/mxnet/module/base_module.py]: eval.data/data_dict  is already sorted( io.py  in  _init_data()) , I am just deleting entries from there, should be already in sorted order.

File: python/mxnet/module/base_module.py [hunk 4]
```diff
@@ -364,6 +385,22 @@ def predict(self, eval_data, num_batch=None, merge_batches=True, reset=True,
 
         output_list = []
 
+        # If NDArrayIter data provided is a dict. This makes sure that only relevant data items
+        # matching the data names, provided during bind time, are send for eval.
+        orig_eval_data = None
+        if isinstance(eval_data, io.NDArrayIter) and \
+                not eval_data.renamed_data and \
+                eval_data.data and \
+                isinstance(eval_data.data[0], tuple):
```
[the remark this answers is not in the document]
Comment (spidyDev) [reply python/mxnet/module/base_module.py]: "eval_data.data"  line 393 checks for that .

File: python/mxnet/io.py [hunk 5]
```diff
@@ -506,6 +507,7 @@ def _init_data(data, allow_empty, default_name):
         else:
             data = OrderedDict( # pylint: disable=redefined-variable-type
                 [('_%d_%s' % (i, default_name), d) for i, d in enumerate(data)])
+        renamed_data = True
```
[the remark this answers is not in the document]
Comment (spidyDev) [reply python/mxnet/io.py]: will do

File: python/mxnet/io.py [hunk 6]
```diff
@@ -645,8 +647,8 @@ def __init__(self, data, label=None, batch_size=1, shuffle=False,
                  label_name='softmax_label'):
         super(NDArrayIter, self).__init__(batch_size)
 
-        self.data = _init_data(data, allow_empty=False, default_name=data_name)
-        self.label = _init_data(label, allow_empty=True, default_name=label_name)
+        self.data, self.renamed_data = _init_data(data, allow_empty=False, default_name=data_name)
+        self.label, _ = _init_data(label, allow_empty=True, default_name=label_name)
```
[the remark this answers is not in the document]
Comment (spidyDev) [reply python/mxnet/io.py]: I was thinking about that too , but was not sure what would be a good place to handle labels, probably at bind . But I have to dig some more.

File: python/mxnet/module/base_module.py [hunk 7]
```diff
@@ -244,6 +245,22 @@ def score(self, eval_data, eval_metric, num_batch=None, batch_end_callback=None,
...
+            # Keeping a copy of original data.
+            orig_eval_data = eval_data.data
+            data_dict = dict(eval_data.data)
+
+            for k, _ in eval_data.data:
+                if k not in self.data_names:
+                    del data_dict[k]
+            eval_data.data = list(data_dict.items())
```
[the remark this answers is not in the document]
Comment (spidyDev) [reply python/mxnet/module/base_module.py]: Did you mean create a  func() ?

File: python/mxnet/module/base_module.py [hunk 8]
```diff
@@ -268,6 +285,10 @@ def score(self, eval_data, eval_metric, num_batch=None, batch_end_callback=None,
             for callback in _as_list(score_end_callback):
                 callback(params)
 
+        # restoring original user data if changed:
+        if orig_eval_data is not None:
```
[the remark this answers is not in the document]
Comment (spidyDev) [reply python/mxnet/module/base_module.py]: hmm .. will have to think about this , coz we cant restore before all the forward op is completed on all the batches. How abt try/catch the forward loop?

File: python/mxnet/io.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (spidyDev) [reply python/mxnet/io.py]: Looking in to this, i do not see a scenario in which we will have issues with the label. User can create  NDIterator for labels and can bind only few of them . It should not affect the forward pass . Correct me if I am missing something here.

File: tests/python/unittest/test_io.py [hunk 9]
```diff
@@ -313,6 +313,20 @@ def check_CSVIter_synthetic():
...
+        'i0': mx.nd.zeros(5, ),
+        'i1': mx.nd.ones(5, ),
+    }, batch_size=5)
+
+    net = mx.sym.var('i1')
+    mod = mx.module.Module(net, ['i1'], None)
+    mod.bind([('i1', (5,))], None)
+    mod.init_params()
+    mod.predict(train_iter)
+    assert_almost_equal(mod.get_outputs()[0].asnumpy(), np.ones(shape=(5,)))
```
Comment (reminisce) [inline tests/python/unittest/test_io.py]: Please add more unit tests to increase test coverage, including assert exception.

**Comment by spidyDev** [comment]
@piiswrong  I agree. But user can create a NDIterator with multiple key-value but bind only few of them. For example,
```
    train_iter = mx.io.NDArrayIter({
        'i2': mx.nd.ones((5, 4, 3, 3)),
        'i0': mx.nd.zeros(5, ),
        'i1': mx.nd.ones(5, ),
        'i3': mx.nd.ones((5, 4, 3, 3))
    }, batch_size=5)

    net = mx.sym.add_n(mx.sym.var('i2'), mx.sym.var('i3'))
    mod = mx.module.Module(net, ['i2', 'i3'], None)
    mod.bind([('i2', (5, 4, 3, 3)), ('i3', (5, 4, 3, 3))])
    mod.init_params()
    mod.predict(train_iter)
    assert_almost_equal(mod.get_outputs()[0].asnumpy(), 2*np.ones(shape=(5, 4, 3, 3)))
```

This might not be ideal , but valid ?

**Comment by piiswrong** [comment]
That's incorrect usage. We shouldn't address it by special casing ndarrayiter in module.

One thing we can do is to check data_iter.provide_data against module(data_names) and raise a warning if they don't match

**Comment by spidyDev** [comment]
@piiswrong  Could you review ?

File: python/mxnet/module/base_module.py [hunk 10]
```diff
@@ -78,6 +79,17 @@ def _parse_data_desc(data_names, label_names, data_shapes, label_shapes):
     return data_shapes, label_shapes
 
 
+def _check_data_names(eval_data, data_names):
+    """ Check if iterator data names match the data names provided in module"""
+    if isinstance(eval_data, io.NDArrayIter) and \
```
Comment (piiswrong) [inline python/mxnet/module/base_module.py]: This is still special casing NDArrayIter. The same problem would happen for all dataiters if names are incorrect.
Why do we need renamed_data attribute?

Comment (spidyDev) [reply python/mxnet/module/base_module.py]: In NDArrayIter, user can provide data in three formats:
data (array or list of array or dict of string to array) 

if the data is not a dict, mxnet will create a dict with built-in data_names, these would not match the data names provided by the user . We need to only match in the scenario if the a dict value was provided by the user while creating NDArrayIter

**Comment by lupesko** [comment]
@spidyDev - what is the next step here?

**Comment by spidyDev** [comment]
Closing this issue, as it may not be relevant at this point.

**Pull request closed without merging**